test/uk/gov/hmrc/gform/models/FormModelBuilderSpec.scala

```diff
@@ -0,0 +1,97 @@
+import uk.gov.hmrc.gform.graph.FormTemplateBuilder._
+import uk.gov.hmrc.gform.models.optics.{ DataOrigin, FormModelVisibilityOptics }
+import uk.gov.hmrc.gform.sharedmodel.SourceOrigin
+import uk.gov.hmrc.gform.sharedmodel.form.FormModelOptics
+import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+
+class FormModelBuilderSpec extends FlatSpec with Matchers with FormModelSupport with VariadicFormDataSupport {
+
+  implicit def implicitToFormComponentId(str: String): FormComponentId = FormComponentId(str)
+
+  "FormModelBuilder.renderPageModel" should "return visibility model without components whose includeIf evaluates to false" in {
+
+    val includeIf = IncludeIf(Equals(FormCtx("a"), Constant("X")))
+
+    val fcA = mkFormComponent("a", Value)
+    val fcB = mkFormComponent("b", Value).copy(includeIf = Some(includeIf))
+    val fcC = mkFormComponent("c", Value)
+
+    val section1 = mkSection(List(fcA))
+    val section2 = mkSection(List(fcB, fcC))
+
+    val sections = List(
+      section1,
+      section2,
+    )
+    val fmb = mkFormModelFromSections(sections)
+
+    val expectedPageA = mkPage(fcA :: Nil)
+
+    val expectedPageB1 = mkPage(fcC :: Nil)
+    val expectedPageB2 = mkPage(fcB :: fcC :: Nil)
+
+    val staticTypeInfo = StaticTypeInfoBuilder.simple(
+      "a" -> ExprType.String,
+      "b" -> ExprType.String,
+      "c" -> ExprType.String,
+    )
+
+    val table = Table(
+      ("data", "expectedData", "expectedPages"),
+      (
+        variadicFormData[SourceOrigin.OutOfDate]("a" -> "NotX", "b" -> "B", "c" -> "C"),
+        variadicFormData[SourceOrigin.Current]("a"   -> "NotX", "b" -> "B", "c" -> "C"),
+        NonEmptyList.of(
+          Bracket.NonRepeatingPage(Singleton(expectedPageA), SectionNumber(0), section1),
+          Bracket.NonRepeatingPage(Singleton(expectedPageB1), SectionNumber(1), section2),
+        )
+      ),
+      (
+        variadicFormData[SourceOrigin.OutOfDate]("a" -> "X", "b" -> "B", "c" -> "C"),
+        variadicFormData[SourceOrigin.Current]("a"   -> "X", "b" -> "B", "c" -> "C"),
+        NonEmptyList.of(
+          Bracket.NonRepeatingPage(Singleton(expectedPageA), SectionNumber(0), section1),
+          Bracket.NonRepeatingPage(Singleton(expectedPageB2), SectionNumber(1), section2),
+        )
+      )
+    )
+
+    forAll(table) {
+      case (data, expectedData, expectedPages) =>
+        val visibilityOptics: FormModelVisibilityOptics[DataOrigin.Mongo] =
+          fmb.visibilityModel[DataOrigin.Mongo, SectionSelectorType.Normal](data, None)
+
+        val formModelOptics: FormModelOptics[DataOrigin.Mongo] =
+          fmb.renderPageModel[DataOrigin.Mongo, SectionSelectorType.Normal](visibilityOptics, None)
+
+        val expected: FormModel[Visibility] = fromPagesWithIndex(expectedPages, staticTypeInfo)
+
+        formModelOptics.formModelVisibilityOptics.formModel shouldBe expected
+        visibilityOptics.recData.variadicFormData shouldBe expectedData
```

Comment:
do we really need the second assertion here?

Reply:
It is there as a sanity check that invisibility of the component doesn't affect data we store for such a component.

Reply:
Ok